client/src/test/java/io/pravega/client/CredentialsExtractorTest.java

```diff
@@ -10,83 +10,202 @@
+        ClientConfig clientConfig = ClientConfig.builder().extractCredentials(properties, null).build();
+        Credentials credentials = clientConfig.getCredentials();
 
-        assertEquals("Method is not picked up from properties",
-                "temp", config.getCredentials().getAuthenticationType());
+        assertNotNull(credentials);
+        assertNotNull("io.pravega.client.ClientConfig$ClientConfigBuilder$1",
+                credentials.getClass());
+        assertEquals("amethod", credentials.getAuthenticationType());
+        assertEquals("atoken", credentials.getAuthenticationToken());
+    }
 
-        assertEquals("Token is not same",
-                "mytoken", config.getCredentials().getAuthenticationToken());
+    @Test
+    public void extractsCredentialsFromEnvVariables() {
+        Map authEnvVariables = new HashMap();
+        authEnvVariables.put("pravega_client_auth_method", "amethod");
+        authEnvVariables.put("pravega_client_auth_token", "atoken");
+
+        ClientConfig clientConfig = ClientConfig.builder().extractCredentials(null, authEnvVariables).build();
+        Credentials credentials = clientConfig.getCredentials();
+
+        assertNotNull(credentials);
+        assertNotNull("io.pravega.client.ClientConfig$ClientConfigBuilder$1",
+                credentials.getClass());
+        assertEquals("amethod", credentials.getAuthenticationType());
+        assertEquals("atoken", credentials.getAuthenticationToken());
+    }
 
-        //If a credential is explicitly mentioned, do not override from properties
-        config = ClientConfig.builder().credentials(new Credentials() {
-            @Override
-            public String getAuthenticationType() {
-                return null;
-            }
+    @Test
+    public void explicitlySpecifiedCredentialsAreNotOverridden() {
+        Properties properties = new Properties();
+        properties.setProperty("pravega.client.auth.method", "amethod");
+        properties.setProperty("pravega.client.auth.token", "atoken");
+
+        Map authEnvVariables = new HashMap();
+        authEnvVariables.put("pravega_client_auth_method", "amethod");
+        authEnvVariables.put("pravega_client_auth_token", "atoken");
+
+        ClientConfig clientConfig = ClientConfig.builder()
+                .credentials(new Credentials() {
+                    @Override
+                    public String getAuthenticationType() {
+                        return "typeSpecifiedViaExplicitObject";
+                    }
+
+                    @Override
+                    public String getAuthenticationToken() {
+                        return "tokenSpecifiedViaExplicitObject";
+                    }
+                }).extractCredentials(properties, authEnvVariables)
+                .build();
+
+        assertEquals("Explicitly set credentials should not be overridden", "typeSpecifiedViaExplicitObject",
```

Comment:
Based on this test case, I´m wondering if we should warn the user that the configuration via system properties and env variables is being ignored. I understand that we follow a given order in the presence of multiple configuration types, but perhaps we want to let the user know that she is not necessarily doing the right thing.

Note that I´m not suggesting a change to the test case, I´m suggesting that we print a warn log message in the case a user specifies multiple types and we by consequence ignore one.

Reply:
Warning the user that the configuration via system properties and environment variables are being ignored would be helpful from the usability perspective. There are a couple of issues with that though:

1. The environment variable or system properties might have been specified by someone else at deployment time while the explicit credentials are specified at development time. The user might be seeing a log message as a consequence of someone else's action. 
2. There is a bit of additional processing that is required to assess the condition, in order to find whether to log the message.

I think we should leave it as is for now, especially because there are adequate log statements that make the condition clear: https://github.com/pravega/pravega/blob/7a9c3595e49cbf192ba713d6863df0a3df44775c/client/src/main/java/io/pravega/client/ClientConfig.java#L123

Would you agree?

Reply:
Let's have info level messages to state what method is being used to load the credentials. That way we avoid any confusion about how we are loading it at run time.

Reply:
Sure, done.